# Sprint 7

Este notebook contiene toda la información sobre el proyecto. Contiene las instrucciones detalladas para completar el proyecto indicadas <a href="https://tripleten.com/trainer/data-scientist/lesson/c779060b-b834-47b3-ac73-7fc21b61558b/">aquí</a>, la explicación, exploración y análisis de los datos, así como la presentación de conclusiones de comentarios sobre el proyecto.

## Instrucciones para completar el proyecto

### Paso 1. Configuración

1. Crea una cuenta en github.com. Si ya tienes una cuenta o creaste una mientras avanzabas en el capítulo Git y GitHub, puedes omitir este paso. ✅

1. Crea un nuevo repositorio git con un archivo README.md y un archivo .gitignore (hay que elegir la plantilla de Python). Si necesitas repasar, consulta esta lección. ✅

1. Si no lo has hecho, crea una cuenta en render.com. <a href="https://tripleten.com/trainer/data-analyst/lesson/c434dc0f-799f-4cfd-a06c-61b49db65fa3/">En esta lección</a> se analizaron las aplicaciones web y de renderizado. Al crear una cuenta de Render, elige la opción "Github" y sigue los pasos de registro. Esto es exactamente lo que necesitas para vincular tu cuenta de Render a tu cuenta de GitHub. ✅

1. Este proyecto implica realizar un análisis exploratorio de datos. Para lograrlo, necesitarás tener instalados los paquetes pandas y plotly-express. Anteriormente analizamos plotly en esta lección. plotly-express está diseñado con valores predeterminados razonables y opciones configurables para tipos de gráficos comunes, lo cual lo convierte en un excelente punto de partida para principiantes en comparación con solo plotly. Si eres principiante en esto, no te preocupes, ¡te guiaremos a través del proceso! ✅

    Además, necesitarás el paquete streamlit para desarrollar una aplicación web.

    Crea un nuevo entorno virtual y asígnale un nombre significativo que esté relacionado con el conjunto de datos con el que trabajarás, por ejemplo, podrías llamarlo vehicles_env.
    Asegúrate de haber instalado al menos los siguientes paquetes en el entorno: <span style="font-family:Courier New">pandas</span>, <span style="font-family:Courier New">plotly_express</span> y <span style="font-family:Courier New">streamlit</span>. ✅ Nombre del entorno *_ssma_env_*.


    1. Si tienes dificultades para encontrarlos, estos son los comandos para instalar las librerías: ✅

        - Para plotly-express, echa un vistazo a <a href="https://anaconda.org/plotly/plotly_express">esta página de documentación</a>.
        - <a href="https://anaconda.org/anaconda/pandas">Esta es la página de doc</a> para pandas donde encontrarás el comando para ejecutar.
        - Por último, <a href="https://anaconda.org/conda-forge/streamlit">esta es una página para streamlit</a>.
        - Además, necesitarás instalar una librería adicional (<span style="font-family:Courier New">nbformat</span>) para habilitar plotly express. Instálala mediante:<p style="font-family:Courier New">conda install -n myenv nbformat</p> 

1. Clona el repositorio de tu proyecto desde GitHub y ábrelo como un proyecto en VS Code. Este será el directorio de tu proyecto. Configura el intérprete de Python como el utilizado por el entorno virtual que creaste anteriormente. ✅

1. Crea el archivo requirements.txt y añade tres librerías ahí sin especificar las versiones: ✅
    <p style="font-family:Courier New">pandas</p>
    <p style="font-family:Courier New">plotly_express</p>
    <p style="font-family:Courier New">streamlit</p>

### Paso 2. Descarga del archivo de datos

1. Descarga el conjunto de datos de anuncios de coches (vehicles_us.csv) o encuentra tu propio dataset en formato CSV. ✅

1. Coloca el conjunto de datos en el directorio del proyecto. ✅

### Paso 3. Análisis exploratorio de datos

1. Crea un directorio llamado notebooks en el directorio de tu proyecto. ✅

1. Crea un Jupyter notebook llamado EDA.ipynb en VS Code y guárdalo en el directorio notebooks de tu proyecto. Recuerda que .ipynb es una extensión de archivo para Jupyter Notebooks. ✅

1. Abre el Jupyter notebook EDA.ipynb y experimenta con <a href="https://plotly.com/python/plotly-express/">plotly-express</a> para crear visualizaciones para el análisis exploratorio básico del conjunto de datos dentro del notebook. ✅

### Paso 4. Desarrollo del cuadro de mandos de la aplicación web

1. Crea un archivo <span style="font-family:Courier New">app.py</span> en la raíz del directorio de tu proyecto. Para crear un archivo .py, haz clic en "New File" (Nuevo archivo) en VS Code y guárdalo en el directorio del proyecto con el nombre deseado y la extensión .py.

1. Importa <span style="font-family:Courier New">streamlit</span> como <span style="font-family:Courier New">st</span>, <span style="font-family:Courier New">pandas</span> como <span style="font-family:Courier New">pd</span> y <span style="font-family:Courier New">plotly_express</span> como <span style="font-family:Courier New">px</span> al principio del archivo.

1. Lee el archivo CSV del conjunto de datos en un DataFrame. El código será el mismo que tenías en el Jupyter Notebook al explorar el conjunto de datos.

1. Ahora, vamos a crear el contenido de nuestra aplicación basada en Streamlit. Esto es lo que queremos que incluyas:

    - Al menos un encabezado (puedes utilizar <span style="font-family:Courier New">st.header()</span> para hacerlo. En la <a href="https://tripleten.com/trainer/data-analyst/lesson/c434dc0f-799f-4cfd-a06c-61b49db65fa3/">lección de aplicaciones web</a> te mostramos cómo crear un encabezado).
    - Crea un botón que, al hacer clic en él, construya un histograma plotly-express. Para hacerlo, considera utilizar las funciones <span style="font-family:Courier New">st.write()</span> y <span style="font-family:Courier New">st.plotly_chart()</span>.

1. Asegúrate de actualizar el archivo README cuando hayas terminado. Este debe incluir una breve descripción del proyecto, donde se explique para qué sirve la aplicación web y qué tipo de funcionalidad proporciona.

1. Recuerda confirmar y empujar todos los cambios a tu repositorio cuando hayas terminado tu trabajo.

Notas a considerar:

- Mientras desarrollas tu aplicación añadiendo un nuevo componente de Streamlit, puedes ejecutar el comando streamlit run app.py desde la terminal para ver el resultado.

- Es una buena práctica confirmar y enviar tu trabajo a un repositorio remoto en GitHub a medida que vas alcanzando ciertos objetivos en el desarrollo de la aplicación (por ejemplo, añades un componente que funciona y la aplicación se ejecuta sin errores). ¡Así que no te olvides de escribir un mensaje de confirmación significativo!

### Paso 5. Despliegue de la versión final de la aplicación en Render

1. Accede a tu cuenta en render.com y crea un nuevo servicio web.

1. Crea un nuevo servicio web enlazado a tu repositorio de Github.

1. Configura el nuevo servicio web Render añadiendo un Build Command que instale todo lo necesario para iniciar tu app, incluyendo streamlit y todos los paquetes de requirements.txt. Utiliza el siguiente comando:<p style="font-family:Courier New">pip install --upgrade pip && pip install -r requirements.txt</p>

1. Añade a tu Start Command: streamlit run app.py. Debería verse así:<p style="font-family:Courier New">streamlit run app.py</p>

1. Despliega en Render y espera a que el build se ejecute con éxito.

1. Comprueba que tu aplicación sea accesible a través de la siguiente URL.

    Si actualizas tu repositorio de GitHub, para implementar la versión más reciente en Render, haz clic en "Manual Deploy" → "Latest Commit" (Implementación manual → Última confirmación).

### Paso 6. Enviar proyecto

Deberás enviar un enlace a tu repositorio de GitHub. Recuerda dejar Público el repositorio. Si queda privado no podremos verlo.

## Análisis exploratorio de datos

In [1]:
import pandas as pd
import plotly.express as px

### Paso 2. Descarga del archivo de datos: Students' Social Media Addiction

#### <a href="https://www.kaggle.com/datasets/adilshamim8/social-media-addiction-vs-relationships">About Dataset</a>

##### Overview
The Student Social Media & Relationships dataset contains anonymized records of students’ social‐media behaviors and related life outcomes. It spans multiple countries and academic levels, focusing on key dimensions such as usage intensity, platform preferences, and relationship dynamics. Each row represents one student’s survey response, offering a cross‐sectional snapshot suitable for statistical analysis and machine‐learning applications.

##### Scope & Coverage
- Population: Students aged 16–25 enrolled in high school, undergraduate, or graduate programs.
- Geography: Multi‐country coverage (e.g., Bangladesh, India, USA, UK, Canada, Australia, Germany, Brazil, Japan, South Korea).
- Timeframe: Data collected via a one‐time online survey administered in Q1 2025.
- Volume: Configurable sample sizes (e.g., 100, 500, 1,000 records) based on research needs.

##### Key variables

<table>
    <thead>
        <tr>
            <th>Variable</th>
            <th>Type</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Student_ID</td>
            <td>Integer</td>
            <td>Unique respondent identifier</td>
        </tr>
        <tr>
            <td>Age</td>
            <td>Integer</td>
            <td>Age in years</td>
        </tr>
        <tr>
            <td>Gender</td>
            <td>Categorical</td>
            <td>“Male” or “Female”</td>
        </tr>
        <tr>
            <td>Academic_Level</td>
            <td>Categorical</td>
            <td>High School / Undergraduate / Graduate</td>
        </tr>
        <tr>
            <td>Country</td>
            <td>Categorical</td>
            <td>Country of residence</td>
        </tr>
        <tr>
            <td>Avg_Daily_Usage_Hours</td>
            <td>Float</td>
            <td>Average hours per day on social media</td>
        </tr>
        <tr>
            <td>Most_Used_Platform</td>
            <td>Categorical</td>
            <td>Instagram, Facebook, TikTok, etc.</td>
        </tr>
        <tr>
            <td>Affects_Academic_Performance</td>
            <td>Boolean</td>
            <td>Self‐reported impact on academics (Yes/No)</td>
        </tr>
        <tr>
            <td>Sleep_Hours_Per_Night</td>
            <td>Float</td>
            <td>Average nightly sleep hours</td>
        </tr>
        <tr>
            <td>Mental_Health_Score</td>
            <td>Integer</td>
            <td>Self‐rated mental health (1 = poor to 10 = excellent)</td>
        </tr>
        <tr>
            <td>Relationship_Status</td>
            <td>Categorical</td>
            <td>Single / In Relationship / Complicated</td>
        </tr>
        <tr>
            <td>Relationship_Status</td>
            <td>Categorical</td>
            <td>Single / In Relationship / Complicated</td>
        </tr>
        <tr>
            <td>Conflicts_Over_Social_Media</td>
            <td>Integer</td>
            <td>Number of relationship conflicts due to social media</td>
        </tr>
        <tr>
            <td>Addicted_Score</td>
            <td>Integer</td>
            <td>Social Media Addiction Score (1 = low to 10 = high)</td>
        </tr>
    </tbody>
<table>

In [2]:
## Carga el dataset en la misma carpeta
df = pd.read_csv('../students_social_media_addiction.csv')

## Pasa todos los strings a minusculas (columnas y datos)
df.columns = df.columns.str.lower()

df['gender'] = df['gender'].str.lower()
df['academic_level'] = df['academic_level'].str.lower()
df['country'] = df['country'].str.lower()
df['most_used_platform'] = df['most_used_platform'].str.lower()
df['affects_academic_performance'] = df['affects_academic_performance'].str.lower()
df['relationship_status'] = df['relationship_status'].str.lower()

## Cambia tipo de datos de 'student_id' a 'str'
df['student_id'] = df['student_id'].astype('str')

## Renombra algunas columnas por versiones mas cortas
colums_name = {
    'avg_daily_usage_hours':'daily_usage',
    'most_used_platform':'platform',
    'affects_academic_performance':'performance',
    'sleep_hours_per_night':'spleep_hours',
    'mental_health_score':'mental_health',
    'relationship_status':'rel_status',
    'conflicts_over_social_media':'conflicts'
    }
df = df.rename( columns=colums_name )

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 705 entries, 0 to 704
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   student_id      705 non-null    object 
 1   age             705 non-null    int64  
 2   gender          705 non-null    object 
 3   academic_level  705 non-null    object 
 4   country         705 non-null    object 
 5   daily_usage     705 non-null    float64
 6   platform        705 non-null    object 
 7   performance     705 non-null    object 
 8   spleep_hours    705 non-null    float64
 9   mental_health   705 non-null    int64  
 10  rel_status      705 non-null    object 
 11  conflicts       705 non-null    int64  
 12  addicted_score  705 non-null    int64  
dtypes: float64(2), int64(4), object(7)
memory usage: 71.7+ KB


Hay un total de 705 entradas. No hay ninguna columna con valores nulos y como se sabe directamente de la fuente, tampoco hay valores duplicados. Solo es necesario cambiar el tipo de dato de 'student_id' a str.

In [4]:
df.sample(5)

student_id  age  gender academic_level      country  daily_usage  \
104        105   22    male       graduate  afghanistan          2.9   
160        161   19  female  undergraduate   bangladesh          5.3   
632        633   23  female       graduate        italy          4.8   
6            7   23    male       graduate      germany          1.5   
323        324   22    male       graduate        india          5.8   

      platform performance  spleep_hours  mental_health       rel_status  \
104   linkedin          no           7.0              7      complicated   
160  instagram         yes           6.1              5           single   
632    twitter          no           7.1              7  in relationship   
6     linkedin          no           8.0              9           single   
323   whatsapp         yes           6.5              5  in relationship   

     conflicts  addicted_score  
104          2               5  
160          3               7  
632          2               5  
6            0               2  
323          4               8

In [5]:
df.describe(include='all')

student_id         age  gender academic_level country  daily_usage  \
count         705  705.000000     705            705     705   705.000000   
unique        705         NaN       2              3     110          NaN   
top             1         NaN  female  undergraduate   india          NaN   
freq            1         NaN     353            353      53          NaN   
mean          NaN   20.659574     NaN            NaN     NaN     4.918723   
std           NaN    1.399217     NaN            NaN     NaN     1.257395   
min           NaN   18.000000     NaN            NaN     NaN     1.500000   
25%           NaN   19.000000     NaN            NaN     NaN     4.100000   
50%           NaN   21.000000     NaN            NaN     NaN     4.800000   
75%           NaN   22.000000     NaN            NaN     NaN     5.800000   
max           NaN   24.000000     NaN            NaN     NaN     8.500000   

         platform performance  spleep_hours  mental_health rel_status  \
count         705         705    705.000000     705.000000        705   
unique         12           2           NaN            NaN          3   
top     instagram         yes           NaN            NaN     single   
freq          249         453           NaN            NaN        384   
mean          NaN         NaN      6.868936       6.226950        NaN   
std           NaN         NaN      1.126848       1.105055        NaN   
min           NaN         NaN      3.800000       4.000000        NaN   
25%           NaN         NaN      6.000000       5.000000        NaN   
50%           NaN         NaN      6.900000       6.000000        NaN   
75%           NaN         NaN      7.700000       7.000000        NaN   
max           NaN         NaN      9.600000       9.000000        NaN   

         conflicts  addicted_score  
count   705.000000      705.000000  
unique         NaN             NaN  
top            NaN             NaN  
freq           NaN             NaN  
mean      2.849645        6.436879  
std       0.957968        1.587165  
min       0.000000        2.000000  
25%       2.000000        5.000000  
50%       3.000000        7.000000  
75%       4.000000        8.000000  
max       5.000000        9.000000

- La edad comprende. de 18 a 24 años. La edad promedio es 20.6 años.

- Hay casi el mismo número de registros de hombres que de mujeres.

- La mayoría no están graduados.

- El país con mas registros es india representando un 16% de usuarios de un total de 110 paises.

- El promedio de uso de redes sociales es de 4.9 horas con una desviación de 1.3 horas de un rango total de 1.5 a 8.5 horas.

- La plataforma más usada es instagram representando un 35% de preferencias de un total de 12 plataformas.

- El 64% de los usuarios considera que afecta en su desempeño académico.

- El promedio de horas de sueño es de 6.9 horas con una desviación de 1.1 horas de un rango total de entre 3.8 a 9.6 horas.

- El promedio de usuarios considera que su salud mental es regular, digamos, ni buena ni mala, pero más buena que mala.

- Más de la mitad de los usuarios son solteros.

- En promedio los usuarios han tenido un total de 2.8 conflictos sociales generados por redes sociales.

- El promedio de estudiantes considera que está enganchado a las redes sociales, pero no lo consideran una adicción grave.

In [6]:
## histograma de edades y generos
fig = px.histogram(df, x="age", color='gender')

fig.show()

- La mayoría de los usuarios tienen entre 19 y 22 años. Esto corresponde con el hecho de que el estudio se hizo sobre estudiantes de universidad que en promedio ingresan a los 18 años y concluyen 4 años después.

- La mayoría de las mujeres encuestadas tienen entre 19 y 20 años, mientras que de hombres tienen entre 21 y 22 años.

In [7]:
## histograma de edades por plataforma
top_platform = df.groupby('platform')['platform'].count().sort_values(ascending=False)[:5]
print('Top 5 plataformas:')
print(top_platform)
fig = px.histogram(df[ df['platform'].isin(top_platform.index) ], x="age", color='platform', facet_col='platform',title='Histograma de edades por plataforma')

fig.show()

Top 5 plataformas:
platform
instagram    249
tiktok       154
facebook     123
whatsapp      54
twitter       30
Name: platform, dtype: int64


- Las plataformas más utilizadas son instagram, tiktok, facebook, whatsapp y twitter.

- Puede notarse que hay preferencia de personas más jovenes por instagram y tiktok.

In [8]:
print(df[ df['platform'].isin(top_platform.index) ].groupby('platform')['daily_usage'].mean().sort_values(ascending=False))

## histograma de uso por plataforma
fig = px.histogram(df[ df['platform'].isin(top_platform.index) ], x="daily_usage", color='platform', facet_col='platform', title='Histograma de uso diario por plataforma')

fig.show()

platform
whatsapp     6.475926
tiktok       5.346104
instagram    4.872289
twitter      4.870000
facebook     4.507317
Name: daily_usage, dtype: float64


- Sobre las plataformas más usadas, en promedio a whatsapp se le dedica más de 6 horas, a tiktok mas de 5 horas y al resto más de 4.

In [9]:
## dispersión de edades y uso diario
fig = px.scatter(df, x="age", y='daily_usage', trendline="ols", trendline_color_override="grey", title='Relación entre edad y uso diario de plataformas')

fig.show()

## medida de correlación
print('Correlación: ', round(df['age'].corr(df['daily_usage']), 2) )

Correlación:  -0.11


- Aunque no es muy significativa, hay una correlación negativa entre la edad de los usuarios y el uso de redes sociales.

In [10]:
## dispersión de uso diario y puntaje de adicción
fig = px.scatter(df, x="daily_usage", y='addicted_score', color='performance', title='Relación entre uso diario y puntaje de adicción')

fig.show()

## Correlación
print('Correlación: ', round(df['daily_usage'].corr(df['addicted_score']), 2) )

Correlación:  0.83


- Existe una clara correlación entre el uso promedio diario de los estudiantes y el puntaje de adicción que se auto asignan.

- Como existe una correlación, aunque baja, entre la edad y uso de plataformas, es de esperarse que exista una correlación entre edad y en nivel de adicción. Esto es, a mayor edad, los estudiantes se perciben menos enganchados a redes sociales. Sin embargo, no es una correlación particularmente alta. Esto se debe a que no cambia mucho el uso diario entre edades.

- Los usuarios que perciben que el uso de redes sociales no ha afectado su desempeño escolar son aquellos que dedican menos horas a las mismas. Aquellos que perciben que han afectado su desempeño, son aquellos que dedican mucho mas horas.

In [11]:
## dispersión de uso diario y horas de sueño por plataforma
fig = px.scatter(df[ df['platform'].isin(top_platform.index) ], x="daily_usage", y='spleep_hours', color='platform', title='Relación entre uso diario y horas de sueño por plataforma')

fig.show()

## Correlación
print('Correlación: ', round(df['daily_usage'].corr(df['spleep_hours']), 2) )

Correlación:  -0.79


- El uso de plataformas implica destinar tiempo a dicha actividad. Como incrementan el tiempo invertido en redes sociales, es de esperarse que afecte el tiempo dedicado a otras áreas. En este caso, se aprecia una fuerte correlación negativa entre el uso diario y las horas de sueño.

In [12]:
print('Promedio de puntuación de salud mental por plataforma')
print(df[ df['platform'].isin(top_platform.index) ].groupby('platform')['mental_health'].mean().sort_values(ascending=False))

## dispersión de uso diario y salud mental por plataforma
fig = px.scatter(df[ df['platform'].isin(top_platform.index) ], x="daily_usage", y='mental_health', color='platform', title='Relación entre uso diario y salud mental plataforma')

fig.show()

## Correlación
print('Correlación: ', round(df['daily_usage'].corr(df['mental_health']), 2) )

Promedio de puntuación de salud mental por plataforma
platform
twitter      6.833333
facebook     6.715447
instagram    6.124498
tiktok       5.714286
whatsapp     5.537037
Name: mental_health, dtype: float64


Correlación:  -0.8


- Como ya se vio, el tiempo invertido en redes sociales como tiktok y whatsapp es más alto que en el resto. Aquí también puede apreciarse.

- Hay una correlación negativa entre el uso de redes sociales y la salud mental autopercibida. Claro que también hay una fuerte correlación entre la salud mental percibida y las horas de sueño.

- Los estudiantes que se perciben con una mejor salud mental son también aquellos que se asignan una puntuación de adicción a redes sociales más baja. Son los que duermen más.

- El estatus de relación no parece afectar ninguna de las variables significativamente.

In [13]:
## dispersión de uso diario y horas de sueño por plataforma
fig = px.scatter(df, x="daily_usage", y='conflicts', color='gender', title='Relación entre uso diario y conflictos generados')

fig.show()

## Correlación
print('Correlación: ', round(df['daily_usage'].corr(df['conflicts']), 2) )

Correlación:  0.8


- Existe una fuerte correlación positiva entre el número de conflictos sociales reportados y el uso diaro dedicado a redes sociales.

- No se percibe relación entre el genero y número de conflictos o cualquier otra variable numérica.

## Concluciones

- El parámetro referencia más importante es el tiempo dedicado a redes sociales o 'daily_usage' medido en horas. Esta cantidad es un valor promedio diario estimado por los usuarios, una referencia de peso que coloca el usuario al uso de redes sociales respecto de otras actividades en el día. No es una cantidad medida rigurosamente o directamente desde algún dispositivo.

- El tiempo dedicado guarda una fuerte correlación con otras variables tales como:
    - Nivel de adicción a redes sociales (positiva, 0.83).
    - Horas de sueño promedio diarias (negativa, -0.79).
    - Salud mental autopercibida (negativa, -0.8).
    - Número de conflictos sociales reportados (positiva, 0.8).

- Parece haber una correlación negativa pequeña (-0.11) entre la edad y el uso de redes sociales. Esto puede deberse a la necesidad del estudiante por acreditar satisfactoriamente sus materias por lo que restringen su uso de redes sociales.